### ISSUE TITLE
Can't sync data from Debezium JSON with decimal.handling.mode = precise

### ISSUE THREAD

**Issue by cyliu0** [risingwavelabs/risingwave#10628]
### Describe the bug

There is a debezium connector configuration `decimal.handling.mode` whose default is `precise`. Now the RisingWave can’t handle it properly if the upstream, like postgres, has a field with type `decimal(m, n)`. Currently, we can change `decimal.handling.mode` to `double` as a workaround. But we might lose precision?

There is a kafka msg example for debezium json. In this case, the `w_tax` is `decimal(4, 4)`

```
"after": {
      "w_id": 1,
      "w_name": "nPfTWhXLN",
      "w_street_1": "VeZCXq271fSJ2riU",
      "w_street_2": "2gDSaARdNTIUDNWN",
      "w_city": "8at9JwUmh4GI55eInA",
      "w_state": "SC",
      "w_zip": "936611111",
      "w_tax": "BAg=",
      "w_ytd": "AcnDgA=="
    },
```

We might have fixed a similar issue for avro via https://github.com/risingwavelabs/risingwave/pull/10368

**Comment by lmatz** [comment]
> But we might lose precision?

Yeah, this is critical for correctness testing when comparing results between RW and PG.

**Comment by hzxa21** [comment]
@cyliu0 What is the schema of the RW table. Can you share the SQL query you use to CREATE TABLE with debezium json?

**Comment by cyliu0** [comment]
@hzxa21 
```sql
-- upstream in PostgreSQL
CREATE TABLE IF NOT EXISTS warehouse (
	w_id INT NOT NULL,
	w_name VARCHAR(10),
	w_street_1 VARCHAR(20),
	w_street_2 VARCHAR(20),
	w_city VARCHAR(20),
	w_state CHAR(2),
	w_zip CHAR(9),
	w_tax DECIMAL(4, 4),
	w_ytd DECIMAL(12, 2),
	PRIMARY KEY (w_id) /*T![clustered_index] CLUSTERED */
)
-- source table in RisingWave
create table warehouse (
      w_id INT PRIMARY KEY,
      w_name VARCHAR,
      w_street_1 VARCHAR,
      w_street_2 VARCHAR,
      w_city VARCHAR,
      w_state VARCHAR,
      w_zip VARCHAR,
      w_tax NUMERIC,
      w_ytd NUMERIC
    )
    with (
      connector = 'kafka',
      topic = 'postgres.public.warehouse',
      kafka.brokers = '${BENCHMARK_KAFKA_NAME}.${BENCHMARK_NAMESPACE}:9092',
      scan.startup.mode = 'earliest'
    ) row format debezium_json;
```

**Comment by StrikeW** [comment]
Can we require users to set `decimal.handling.mode = string` as a solution for loss precision? The crate we are using support `Decimal::from_str` which can respect precision.

**Comment by hzxa21** [comment]
> decimal.handling.mode

Good idea. Can we have a try? @cyliu0

**Comment by tabVersion** [comment]
> Can we require users to set decimal.handling.mode = string as a solution for loss precision?

I suspect limiting users' settings, as prev we restrict users to set debezium replica mode to `full` in pg, but in production, no one will do that.
All we can do is try our best to accommodate both formats and warn users about the possibility of losing precision because we accept decimal with at most 28 digits, while in pg it is up to 38 digits.

**Comment by hzxa21** [comment]
> All we can do is try our best to accommodate both formats and warn users about the possibility of losing precision because we accept decimal with at most 28 digits, while in pg it is up to 38 digits

+1. It is hard to fullfill all user needs especially considering that we need to support data types from different upstream source.

However, I think the motivation of this issue is to figure out a way to unblock the correctness test so that we can compare the output of RW and PG with some level of confidence. So setting `decimal.handling.mode = string` maybe the simpliest workaround for the test.

I am still curious on why we will lost precision for `decimal(4, 4)` though. Is the precision already lost in the kafka message?

**Comment by tabVersion** [comment]
> I am still curious on why we will lost precision for decimal(4, 4) though. Is the precision already lost in the kafka message?

I am working on that. I will post the json content in the thread, both dbz settings and message key & payload

**Comment by cyliu0** [comment]
> Can we require users to set `decimal.handling.mode = string` as a solution for loss precision? The crate we are using support `Decimal::from_str` which can respect precision.

How can I use this in create table/source on RW? Any ref?

**Comment by cyliu0** [comment]
> I am still curious on why we will lost precision for `decimal(4, 4)` though. Is the precision already lost in the kafka message?

@hzxa21 I'm not saying we already lost precision for this specific type. I think there will be precision loss according to the doc if we use `decimal.handling.mode=double` for some decimal types. https://debezium.io/documentation/reference/stable/connectors/mysql.html#mysql-property-decimal-handling-mode
```
double represents them using double values, which may result in a loss of precision but is easier to use.
```

**Comment by wugouzi** [comment]
reference: https://debezium.io/documentation/faq/#how_to_retrieve_decimal_field_from_binary_representation

**Comment by tabVersion** [comment]
value = 20.50

---

for `decimal` 

scheme 

```
{"type":"struct","fields":[{"type":"int32","optional":false,"field":"scale"},{"type":"bytes","optional":false,"field":"value"}],"optional":true,"name":"io.debezium.data.VariableScaleDecimal","version":1,"doc":"Variable scaled decimal","field":"price"}
```

payload

```
"price":{"scale":1,"value":"AM0="}
```

---

for `decimal(4, 2)`

scheme 

```
{"type":"bytes","optional":true,"name":"org.apache.kafka.connect.data.Decimal","version":1,"parameters":{"scale":"2","connect.decimal.precision":"4"},"field":"price"}
```

payload

```
"price":"CAI="
```

**Comment by tabVersion** [comment]
> > Can we require users to set `decimal.handling.mode = string` as a solution for loss precision? The crate we are using support `Decimal::from_str` which can respect precision.
> 
> How can I use this in create table/source on RW? Any ref?

I think @hzxa21 is indicating the debezium setting

**Comment by StrikeW** [comment]
> How can I use this in create table/source on RW? Any ref?

You can refer to `integration_tests/debezium-postgres/postgres/postgres_dbz.sh` in the risngwave repo. @cyliu0

**Comment by xiangjinwu** [comment]
The `io.debezium.data.VariableScaleDecimal` support was wrong: https://github.com/risingwavelabs/risingwave/pull/10792#discussion_r1879748452

The fixed decimal case remain unsupported until we build the Kafka Connect layer: #16852
